### ISSUE TITLE
Improve documentation for uploading files to Sharepoint via Graph

### ISSUE THREAD

**Issue by mmarcotte** [microsoftgraph/microsoft-graph-docs#18068]
Your docs for posting files to a Sharepoint page specify:

```
PUT /sites/{site-id}/drive/items/{parent-id}:/{filename}:/content
```

However, I found that confusing for a few reasons.

1. `{parent-id}` is not documented elsewhere on this page. It could be a resource? Or name of the folder within which to put the file? 
2. I could not get `/items/` to work, and could only get `/root` to work after `/drive`.
3. If the file is to be uploaded into a directory, that is included in the filename. 

For example, I was able to upload into a directory using the following api call 

`/sites/${siteId}/drive/root:${folderName}/${filename}:/content`

`folderName` can be blank or it must be prefixed with a slash

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: d92fb8c7-4409-d945-57dd-5d6fc4010051
* Version Independent ID: f77c05bd-2688-bd4e-1f8f-f18feb3a8639
* Content: [Upload small files - Microsoft Graph v1.0](https://docs.microsoft.com/en-us/graph/api/driveitem-put-content?view=graph-rest-1.0&tabs=http)
* Content Source: [api-reference/v1.0/api/driveitem-put-content.md](https://github.com/microsoftgraph/microsoft-graph-docs/blob/main/api-reference/v1.0/api/driveitem-put-content.md)
* Product: **sharepoint**
* Technology: **microsoft-graph**
* GitHub Login: @JeremyKelley
* Microsoft Alias: **MSGraphDocsVteam**

**Comment by Dean-NC** [comment]
I agree that `PUT /sites/{site-id}/drive/items/{parent-id}:/{filename}:/content` needs investigation. I had trouble with it also. I don't know if `{parent-id}` is the Id of a drive or folder.

**Comment by greatBigMassive** [comment]
Hello,

Using PHP here, browser based (HTML5, Vanilla JavaScript)

In the upload sections of the documentation there is nothing (that I can see) that explains the Business OneDrive uploadUrl returned by the Graph API as it is much different to all of the documentation.  Example : `https://{tenant}.sharepoint.com/personal/{ME}/_api/v2.0/drive/items/{folder_id}/uploadSession?guid='{this_has_single_quotes_around_it_and_is_confusing}'&overwrite=True&rename=False&dc=0&tempauth={very_long_string}`

Also, nothing, anywhere explains how you are supposed to allow a client on a Browser (PC using Edge) to PUT their upload into the uploadUrl.  Everyone keeps prefixing "PUT" in front of the Url and assuming everyone else knows what the conversion of that is for a browser.    I have tried `<form method='put' etc >` and could only get it to submit if I removed all the params from the uploadUrl and then inserted them as hidden fields but in all honesty, I have no idea if these is even the correct way to do it. Nothing appears to exist to explain how a user on a client machine uploads a file directly from their machine to the uploadUrl provided using a form or vanilla javascript.    Do you use HTML5 FileReader? Can you use a standard form? (I assume not).

I am pretty sure I don't have to upload the file from the client to my server THEN upload it from my server to OneDrive via PHP cURL as that would be stupid.

Sorry if this is in the wrong place.
Thank you in advance,

**Comment by Dean-NC** [comment]
@greatBigMassive HTML forms only support GET and POST as methods. The JavaScript "fetch API" (or older AJAX or jQuery.ajax()) can make an HTTP request (any HTTP verb...PUT, etc.) to any URL. To upload with fetch (it can use a file input directly), you don't need FileReader API but could use it. Or as you said you could do it server-side in PHP, and I don't consider that stupid. Either way would work.

[https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch)
There's an "Uploading a file" section on that page that uses `FormData` as an example, but the "body" property of the fetch's config property accepts an actual HTML file input as its value.

https://stackoverflow.com/questions/36067767/how-do-i-upload-a-file-with-the-js-fetch-api
See accepted answer.

**Comment by greatBigMassive** [comment]
@Dean-NC Thanks Dean, I have got a bit further on it actually. It's a massive learning curve but your links have helped! 👍 

I used FormData() and XMLHttpRequest() with some Content-Range stuff about an hour ago and now I am at least getting some JSON responses with chunked bytes. I appreciate this isn't a support forum but I've noticed many people getting the same issue I am so might be something to detail more in documentation? ` {"code":"invalidRequest","message":"The Content-Range header length does not match the provided number of bytes."}}`

I decided to try a basic text file instead of a JPG image and created a test.txt file 56 bytes long.
It absolutely uploads the chunks and because it's so small it shows the Blob size is 56 and I pass the header `Content-Range: bytes 0-55/56`  ( I don't even need to chunk it) but it always returns the match issue. If I try change it in anyway (even the number from 0-55 to 0-56)  I get a malformed message instead.   No really seen a solution.
Not expecting support but letting you know as it seems to be quite common but with no real answers.

Thank you and have a good evening.
I'd say my original request/question has been answered but not sure how this is closed as a result.

**Comment by Dean-NC** [comment]
@greatBigMassive There's also this article if you need chunking, but since Graph API has a dedicated "upload session" for large files (article I referenced earlier), Graph may not work with chunking unless you use it's way.

https://dev.to/api_video/uploading-large-files-with-javascript-o4i

**Comment by greatBigMassive** [comment]
@Dean-NC Thank you for the examples and the link - that's been really helpful.
The chunking seems to work fine with JS (The resumable uploads method) so going to leave it that way but like you said, I'll check the file size and then only chunk if it's big and go with your method.

Thank you, have a good day!
